sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala

```diff
@@ -203,6 +203,21 @@ trait FileFormat {
    * method. Technically, a file format could choose suppress them, but that is not recommended.
    */
   def metadataSchemaFields: Seq[StructField] = FileFormat.BASE_METADATA_FIELDS
+
+  /**
+   * The extractors to use when deriving file-constant metadata columns for this file format.
+   *
+   * By default, the value of a file-constant metadata column is obtained by looking up the column's
+   * name in the file's metadata column value map. However, implementations can override this method
+   * in order to provide an extractor that has access to the entire [[PartitionedFile]] when
+   * deriving the column's value.
+   *
+   * NOTE: Extractors are lazy, invoked only if the query actually selects their column at runtime.
+   *
+   * See also [[FileFormat.getFileConstantMetadataColumnValue]].
+   */
+  def fileConstantMetadataExtractors: Map[String, PartitionedFile => Any] =
```

Comment:
I have 2 questions about the API design:
1. File source implementations now need to define constant metadata columns in both `metadataSchemaFields` and `fileConstantMetadataExtractors`. Shall we have a single API to define constant metadata columns? The API should define both column name, type, and extractor.
2. Should the extractor return a `Literal` instead of `Any`?

Reply:
> Shall we have a single API to define constant metadata columns?

Maybe? I couldn't figure out a good way to do it:

> I originally wanted to just embed the extractor in the attribute definition directly, but attribute/struct metadata cannot store code and IMO that's a wise design decision we shouldn't change

I guess we could store pairs or a triple, but (a) the extractors are optional (we'd be forcing every column to either use one or opt out); and (b) the uses for the attribute/struct definitions are rather separated from the uses of the extractors (result: a lot of `x.map` and `x.filter` type calls clutter up the code)

> Should the extractor return a `Literal` instead of `Any`?

I thought about it? But:
* it complicates the definitions of all extractors with boilerplate, which IMO indicates a flawed design
* we anyway have to wrap it in a literal again to handle the null case (or require every extractor to think about nulls). 
* Currently, `getFileConstantMetadataColumnValue` takes care of converting to literal, and that's what everyone uses (nobody calls the extractor directly).
* It would be non-symmetric with the default case, that pulls `Any` from the column value map.

Note that nothing prevents an extractor from returning a literal, because `Literal.apply` will detect and return it unchanged.

Overall, I worry the above changes would only give the code a bigger surface area with little/no gain in safety or functionality?

Reply:
Can we have 2 APIs?
```
def constantMetadataCols: Seq[(StructField, PartitionedFile => Any)] = ...

def generatedMetadataCols: Seq[StructField]
```

Reply:
My earlier response assumed you meant two APIs, and was pointing out that it would cause headaches for both user and implementation because most metadata columns don't need extractors, and the code that uses extractors is different from the code that uses the attributes.

Reply:
Meanwhile, what "safety" issue are we even responding to here? Right now, the behavior of a file-constant metadata column is perfectly predictable (and consistent with its behavior before this PR) -- _unless_ the user opts in to using an extractor by overriding the method to provide an extractor for it.

Reply:
Today's API is two pieces:
```scala
/** All fields of the _metadata struct */
def metadataSchemaFields: Seq[StructField]
/** Custom extractors for the subset of file-constant metadata columns that actually need them */
def fileConstantMetadataExtractors: Map[String, PartitionedFile => Any]
```
Splitting the API would produce instead (at a minimum):
```scala
/** Extractor for file-constant metadata columns that just pull the value from the metadata map */
val DEFAULT_EXTRACTOR: PartitionedFile => Any
/** Defines the file-constant fields of the _metadata struct */
def constantMetadataCols: Seq[(StructField, PartitionedFile => Any)]
/** Defines the file-generated fields of the _metadata struct */
def generatedMetadataCols: Seq[StructField]
```
The only safety issue I thought of so far, is that it might be possible to get the name mapping wrong between columns and their extractors. But if that happened even the most basic unit test for the new column would fail. Meanwhile, that mapping is anyway needed internally because the list of metadata columns is named attributes or even column names, so the pairing proposed above would anyway have to be reconstructed.

Reply:
I just realized a potential difference of underlying assumptions that could be making this discussion more difficult:

The reviewer comments so far are focused on what the changes in this PR look like. However, this PR isn't representative of the experience somebody would have while trying to define custom metadata columns for their file format. By definition all of the standard columns `FileFormat` natively provides are _NOT_ custom and come directly from the fields of `PartitionedFile` itself, rather than indirectly from the metadata values map which users have control over.

My thought process for designing the API was more focused on ease of use for the former case (custom file formats), since the latter (base file format) is a one-time effort within in this PR.

Reply:
My preference is to have a simple and clean framework to generate constant metadata columns. I think the framework is a bit complicated right now:
1. FileFormat implementations can define a few constant metadata columns (name and data type).
2. The constant metadata columns should either match a built-in name in `FileScanRDD`, or be filled by a custom `FileIndex`. With this PR, there is no built-in name. It's either filled by `FileIndex` or has an extractor.
3. With this PR, FileFormat implementations can provide extractors for certain constant metadata columns.

Basically, we get the value of a constant metadata column from two maps: one map is filled by `FileIndex` and contains the value directly. The other map is filled by the `FileFormat` implementation and contains extractors.

It will be great if we can replace the first map with the second one. But if we can't, I'm fine with the current API as we can't avoid name lookup anyway. Can we add more comments in the new API, saying that it may only handle part of the constant metadata columns? The rest of them should be handled by a custom `FileIndex`.

Reply:
I made a dummy PR https://github.com/apache/spark/pull/40930 to show what the other approach actually looks like. I still think it is uglier code for everyone involved, with no benefit to safety, but at least this way we have something concrete to compare.

The change for the other PR vs master is:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala                              |  33 +++++++++++++---------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 145 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala                         |  70 +++++++++++++++--------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala          |  10 ++++-----
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala           |   4 ++--
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala |  71 ++++++++++++++++++++++++++++++++++++++++++++++++------------
 6 files changed, 196 insertions(+), 137 deletions(-)
```

The change of this PR vs. master touches fewer files and is ~50LoC smaller:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala                              |  33 +++++++++++++++------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 102 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala                         |  70 +++++++++++++++++-----------------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala          |  10 +++++-----
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala |  35 ++++++++++++++++++++++++++++++++++-
 5 files changed, 143 insertions(+), 107 deletions(-)
```

The other PR vs. this one is changes 3 files and ~100LoC:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 99 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala           |  4 ++--
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala | 70 ++++++++++++++++++++++++++++++++++++++++------------------------------
 3 files changed, 98 insertions(+), 75 deletions(-)
```

Reply:
> Basically, we get the value of a constant metadata column from two maps: one map is filled by FileIndex and contains the value directly. The other map is filled by the FileFormat implementation and contains extractors.

Nit: The extractors don't "contain" the value at all. They just tell how to extract a value from the `PartitionedFile` that the `FileIndex` provides. I don't know a cleaner way to do this, given that the extraction policy should probably be consistent across all files, rather than re-defined by each file? (prior to this PR, the "policy" was hard-wired as "look it up in the map"; now it's allowed to also be "do arbitrary computation on `PartitionedFile`, if the user prefers)

Reply:
> Can we add more comments in the new API, saying that it may only handle part of the constant metadata columns? The rest of them should be handled by a custom FileIndex.

There was already one comment to that effect in the PR that added custom metadata column support:
https://github.com/apache/spark/pull/40677/files#diff-2d7ec6604810e1be58ca7852ec389853c143b684894f2390ffe07f96701d2219R29

And the unit tests also are a pretty clear demonstration. 

Good point that the comments are buried pretty deep tho -- I'll try to add/expand them in FileFormat.scala as well.

Reply:
doc comments updated, PTAL